airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx

```diff
@@ -65,7 +55,7 @@ export const ConnectorBuilderStateProvider: React.FC<React.PropsWithChildren<unk
     DEFAULT_BUILDER_FORM_VALUES
   );
   const builderFormValues = useMemo(() => {
-    return { ...DEFAULT_BUILDER_FORM_VALUES, ...(storedBuilderFormValues ?? {}) };
+    return merge(DEFAULT_BUILDER_FORM_VALUES, storedBuilderFormValues);
```

Comment:
I added this change because I realized that using the spread operator causes the top-level properties to be completely overwritten. This was a problem for me because my locally stored values did not have any `global.authenticator` set, so I loaded up this branch, the entire locally-stored `global` value overrwrote the default `global` value, meaning that `global.authenticator` was no longer set and some code that was trying to access that failed, so the webapp crashed.

I therefore switched to using lodash `merge` here, so that sub-properties in the default values would still remain if they didnt exist in the locally store values. However, when I added this it caused this entire state provider to seemingly stop re-rendering as a result of the builderFormValues changing, even though there is a useEffect below that consumes it.

This meant that whenever I changed a value in the forms in the UI, that change was not being reflected in the side panel, because the `jsonManifest` values wasn't actually changing anymore. 

@flash1293 do you have any idea why switching to lodash `merge()` here would cause that behavior? I tried logging out the `builderFormValues` and I confirmed that they were indeed changing as expected, but for some reason the useEffect hook that is dependent on it stopped firing.

Reply:
React is doing change detection purely based on the reference of the object you pass in. So if you mutate an object, its reference will stay the same and React will not run effect hooks that depend on it or re-render memoized components if its passed in as a prop.

As you can see here https://lodash.com/docs/4.17.15#merge , lodash merge is mutating the object you pass in as the first argument, merging in all the props from the tail of arguments. This is very much not what you want as it will mutate `DEFAULT_BUILDER_FORM_VALUES`.

You can solve this by creating a new object to merge everything into as part of the useMemo hook:
```
merge({}, DEFAULT_BUILDER_FORM_VALUES, storedBuilderFormValues);
```

As `{}` will create a new object, `builderFormValues` will change and effect hooks and memoized components will re-evaluate.